src/net_processing.cpp

```diff
@@ -3988,6 +3988,55 @@ void PeerLogicValidation::ConsiderEviction(CNode& pto, int64_t time_in_seconds)
 
 void PeerLogicValidation::EvictExtraOutboundPeers(int64_t time_in_seconds)
 {
+    // If we have any extra block-relay-only peers, disconnect the youngest unless
+    // it's given us a block -- in which case, disconnect the second youngest
+    // instead (so we rotate our youngest block-relay-only peer).
+    // The youngest block-relay-only peer would be the extra peer we connected
+    // to temporarily in order to sync our tip; see net.cpp.
+    int extra_block_relay_peers = m_connman.GetExtraBlockRelayCount();
+    if (extra_block_relay_peers > 0) {
+        // Pick the block-relay-only peer that we just connected, unless it gave us
+        // a block -- in which case disconnect the next-youngest block-relay-only
+        // peer. We use higher nodeid as a measure for more recent connection.
+        std::pair<NodeId, int64_t> youngest_peer = {-1, 0};
+        std::pair<NodeId, int64_t> next_youngest_peer = {-1, 0};
+
+        m_connman.ForEachNode([&](CNode* pnode) {
+            LockAssertion lock(::cs_main);
+            if (!pnode->IsBlockOnlyConn() || pnode->fDisconnect) return;
+            if (pnode->GetId() > youngest_peer.first) {
+                next_youngest_peer = youngest_peer;
+                youngest_peer.first = pnode->GetId();
+                youngest_peer.second = pnode->nLastBlockTime;
+            }
+        });
+        NodeId to_disconnect = youngest_peer.first;
+        if (youngest_peer.second > next_youngest_peer.second) {
+            // Our newest block-relay-only peer gave us a block more recently;
+            // disconnect our second youngest.
+            to_disconnect = next_youngest_peer.first;
+        }
+        m_connman.ForNode(to_disconnect, [&](CNode* pnode) {
+            LockAssertion lock(::cs_main);
+            // Make sure we're not getting a block right now, and that
+            // we've been connected long enough for this eviction to happen
+            // at all.
+            // Note that we only request blocks from a peer if we learn of a
+            // valid headers chain with at least as much work as our tip.
```

Comment:
"so any in-flight block is a good hint that such peer will soon provide us with useful work"

Reply:
This seems self-evident!  Leaving as-is.